In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("clean_data.csv")
df.head()

,User's ID,ProdID,Rating,Review Count,Category,Brand,Name,ImageURL,Description,Tags
0,1705.0,2.0,0.0,0.0,"premium, beauty, premium, makeup, premium, nai...",opi,"OPI Infinite Shine, Nail Lacquer Nail Polish, ...",https://i5.walmartimages.com/asr/0e1f4c51-c1a4...,NaN,"premium, beauty, premium, makeup, premium, nai..."
1,95.0,76.0,0.0,0.0,"beauty, hair, care, hair, color, auburn, hair,...",easy,"Nice n Easy Permanent Color, 111 Natural Mediu...",https://i5.walmartimages.com/asr/9c8e42e4-13a5...,"pack, 3, pack, 3, upc, 381519000201, beautiful...","beauty, hair, care, hair, color, auburn, hair,..."
2,8.0,8.0,4.5,29221.0,"beauty, hair, care, hair, color, permanent, ha...",clairol,Clairol Nice N Easy Permanent Color 7/106A Nat...,https://i5.walmartimages.com/asr/e3a601c2-6a2b...,"clairol, nice, n, easy, permanent, color, give...","beauty, hair, care, hair, color, permanent, ha..."
3,4.0,3.0,0.0,0.0,"beauty, makeup, lip","kokie, cosmetics","Kokie Professional Matte Lipstick, Hot Berry, ...",https://i5.walmartimages.com/asr/25b4b467-bc61...,"calling, matte, lip, lovers, indulge, rich, cr...","beauty, makeup, lip, kokie, cosmetics, calling..."
4,990.0,3.0,0.0,131.0,"seasonal, stock, essentials, personal, care, s...",gillette,"Gillette TRAC II Plus Razor Blade Refills, Fit...",https://i5.walmartimages.com/asr/1a2ebb06-cd01...,"1971, gillette, introduced, trac, ii, razor, s...","seasonal, stock, essentials, personal, care, s..."


In [ ]:
df.columns
df.shape

(4091, 10)

In [ ]:
print(df.columns)

Index(['User's ID', 'ProdID', 'Rating', 'Review Count', 'Category', 'Brand',
       'Name', 'ImageURL', 'Description', 'Tags'],
      dtype='object')


In [ ]:
df.isnull().sum()

,0
User's ID,1
ProdID,1
Rating,1
Review Count,1
Category,10
Brand,97
Name,1
ImageURL,1
Description,918
Tags,2


In [ ]:
df.dropna(subset=["User's ID", "ProdID", "Rating"], inplace=True)

text_columns = ["Category", "Brand", "Description", "Tags"]
df[text_columns] = df[text_columns].fillna("")

df.reset_index(drop=True, inplace=True)

In [ ]:
df.isnull().sum()

,0
User's ID,0
ProdID,0
Rating,0
Review Count,0
Category,0
Brand,0
Name,0
ImageURL,0
Description,0
Tags,0


In [ ]:
df["User's ID"] = pd.to_numeric(df["User's ID"], errors='coerce')
df["ProdID"] = pd.to_numeric(df["ProdID"], errors='coerce')
df = df[(df["User's ID"] != 0) & (df["ProdID"] != 0)]
df.reset_index(drop=True, inplace=True)

In [ ]:
df.head()

,User's ID,ProdID,Rating,Review Count,Category,Brand,Name,ImageURL,Description,Tags
0,1705.0,2.0,0.0,0.0,"premium, beauty, premium, makeup, premium, nai...",opi,"OPI Infinite Shine, Nail Lacquer Nail Polish, ...",https://i5.walmartimages.com/asr/0e1f4c51-c1a4...,,"premium, beauty, premium, makeup, premium, nai..."
1,95.0,76.0,0.0,0.0,"beauty, hair, care, hair, color, auburn, hair,...",easy,"Nice n Easy Permanent Color, 111 Natural Mediu...",https://i5.walmartimages.com/asr/9c8e42e4-13a5...,"pack, 3, pack, 3, upc, 381519000201, beautiful...","beauty, hair, care, hair, color, auburn, hair,..."
2,8.0,8.0,4.5,29221.0,"beauty, hair, care, hair, color, permanent, ha...",clairol,Clairol Nice N Easy Permanent Color 7/106A Nat...,https://i5.walmartimages.com/asr/e3a601c2-6a2b...,"clairol, nice, n, easy, permanent, color, give...","beauty, hair, care, hair, color, permanent, ha..."
3,4.0,3.0,0.0,0.0,"beauty, makeup, lip","kokie, cosmetics","Kokie Professional Matte Lipstick, Hot Berry, ...",https://i5.walmartimages.com/asr/25b4b467-bc61...,"calling, matte, lip, lovers, indulge, rich, cr...","beauty, makeup, lip, kokie, cosmetics, calling..."
4,990.0,3.0,0.0,131.0,"seasonal, stock, essentials, personal, care, s...",gillette,"Gillette TRAC II Plus Razor Blade Refills, Fit...",https://i5.walmartimages.com/asr/1a2ebb06-cd01...,"1971, gillette, introduced, trac, ii, razor, s...","seasonal, stock, essentials, personal, care, s..."


In [ ]:
df.shape

(3841, 10)

In [ ]:
user_item_matrix = df.pivot_table(
    index="User's ID",
    columns="ProdID",
    values="Rating",
    fill_value=0
)

In [ ]:
user_item_matrix.head()

ProdID,-2.147484e+09,1.000000e+00,2.000000e+00,3.000000e+00,4.000000e+00,5.000000e+00,6.000000e+00,7.000000e+00,8.000000e+00,9.000000e+00,...,7.642779e+08,7.765597e+08,8.093642e+08,8.304885e+08,8.506360e+08,8.685618e+08,9.562188e+08,9.709125e+08,1.513514e+09,2.135948e+09
User's ID,,,,,,,,,,,,,,,,,,,,,
-2.147484e+09,0.000000,5.0,1.74,0.0,0.0,2.3,0.0,0.0,3.6,2.25,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1.000000e+00,3.342857,0.0,5.00,4.5,0.0,0.0,4.2,4.5,0.0,0.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2.000000e+00,2.122222,4.6,3.30,4.0,5.0,4.5,0.0,4.0,0.0,0.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3.000000e+00,1.175000,4.5,0.00,3.9,0.0,4.5,0.0,0.0,0.0,3.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4.000000e+00,2.600000,0.0,4.50,0.0,0.0,4.4,0.0,5.0,0.0,4.80,...,0.0,0.0,0.0,4.0,0.0,0.0,0.0,0.0,0.0,0.0


In [ ]:
user_item_matrix.shape

(1612, 1555)

**The above means 1612 unique users and 1555 unique products**

In [ ]:
df.to_csv("clean_data.csv", index=False)

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity
user_similarity = cosine_similarity(user_item_matrix)

In [ ]:
user_item_matrix.to_csv("user_item_matrix.csv")